modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/ImageMosaicReader.java

```diff
@@ -385,7 +386,18 @@ private void initReaderFromURL(final Object source, final Hints hints) throws Ex
                             ImageMosaicConfigHandler.createGranuleCatalogFromDatastore(
                                     parentDirectory, datastoreProperties, true, getHints());
                 } else {
-                    CatalogConfigurationBean bean = beans.get(0).getCatalogConfigurationBean();
+                    // We need to figure out if there is an heterogeneous mosaic in the list and use
+                    // it as reference. (Otherwise, read on heterogeneous coverages may not work
+                    // properly)
+                    List<MosaicConfigurationBean> referenceBeans =
```

Comment:
You actually just want the first no? Then it's slightly better to use findFirst and then check the returned Optional, and use the "orElse" to pick the first from the original list as a fallback.

Reply:
Oki, Will do.
Thanks for the feedback